File: GRMCore/Class/cSetSoilDepth.vb
Public Class cSetSoilDepth
    Public Enum SoilDepthCode
        D
        M
        S
        VD
        VS
        USER
        CONSTV
        NULL
    End Enum

    Public mSoilDepthDataType As Nullable(Of cGRM.FileOrConst)
    Public mGridSoilDepthFPN As String
    Public mSoilDepthVATFPN As String
    Public mdtSoilDepthInfo As GRMProject.SoilDepthDataTable
    Public mConstSoilDepth As Nullable(Of Single)

    Public Sub New()

    End Sub

    Public Sub SetValues(ByVal prjdb As GRMProject)
        Dim row As GRMProject.ProjectSettingsRow = CType(prjdb.ProjectSettings.Rows(0), GRMProject.ProjectSettingsRow)
        If mSoilDepthDataType.HasValue Then
            With row
                .SoilDepthDataType = mSoilDepthDataType.Value.ToString
                Select Case mSoilDepthDataType.Value
                    Case cGRM.FileOrConst.File
                        .SoilDepthFile = mGridSoilDepthFPN
                        .SoilDepthVATFile = mSoilDepthVATFPN
                        .SetConstantSoilDepthNull()
                    Case cGRM.FileOrConst.Constant
                        .SetSoilDepthFileNull()
                        .SetSoilDepthVATFileNull()
                        .ConstantSoilDepth = CStr(mConstSoilDepth.Value)
                End Select
            End With
        End If
        If mSoilDepthDataType.Equals(cGRM.FileOrConst.File) Then
            mdtSoilDepthInfo.AcceptChanges()
            For Each r As DataRow In mdtSoilDepthInfo.Rows
                r.SetAdded()
            Next
        End If

    End Sub

    Public Sub GetValues(ByVal prjdb As GRMProject)
        mSoilDepthDataType = Nothing
        mSoilDepthVATFPN = Nothing
        mConstSoilDepth = Nothing
        Dim row As GRMProject.ProjectSettingsRow = CType(prjdb.ProjectSettings.Rows(0), GRMProject.ProjectSettingsRow)
        With row
            If Not .IsSoilDepthDataTypeNull Then
                Select Case .SoilDepthDataType
                    Case cGRM.FileOrConst.File.ToString
                        mSoilDepthDataType = cGRM.FileOrConst.File
                        mGridSoilDepthFPN = .SoilDepthFile
                        mSoilDepthVATFPN = .SoilDepthVATFile
                    Case cGRM.FileOrConst.Constant.ToString
                        mSoilDepthDataType = cGRM.FileOrConst.Constant
                        mConstSoilDepth = CSng(.ConstantSoilDepth)
                    Case Else
                        Throw New InvalidOperationException
                End Select
            End If
        End With
        If mSoilDepthDataType.Equals(cGRM.FileOrConst.File) Then
            mdtSoilDepthInfo = New GRMProject.SoilDepthDataTable
            mdtSoilDepthInfo = prjdb.SoilDepth
        Else
            mdtSoilDepthInfo = Nothing
        End If

    End Sub

    Public ReadOnly Property IsSet() As Boolean
        Get
            Return mSoilDepthDataType.HasValue
        End Get
    End Property

    Public Shared Function GetSoilDepthCode(ByVal inName As String) As SoilDepthCode
        Select Case Trim(inName)
            Case SoilDepthCode.D.ToString : Return SoilDepthCode.D
            Case SoilDepthCode.M.ToString : Return SoilDepthCode.M
            Case SoilDepthCode.S.ToString : Return SoilDepthCode.S
            Case SoilDepthCode.VD.ToString : Return SoilDepthCode.VD
            Case SoilDepthCode.VS.ToString : Return SoilDepthCode.VS
            Case SoilDepthCode.USER.ToString : Return SoilDepthCode.USER
            Case SoilDepthCode.CONSTV.ToString : Return SoilDepthCode.CONSTV
            Case SoilDepthCode.NULL.ToString : Return SoilDepthCode.NULL
            Case Else : Return Nothing
        End Select
    End Function
End Class

File: GRM_CSharp/GRMCore/Class/cSetSoilDepth.cs
using System;
using System.Data;

namespace GRMCore
{
    public class cSetSoilDepth
    {
        public enum SoilDepthCode
        {
            D,
            M,
            S,
            VD,
            VS,
            USER,
            CONSTV,
            NULL
        }

        public Nullable<cGRM.FileOrConst> mSoilDepthDataType;
        public string mGridSoilDepthFPN;
        public string mSoilDepthVATFPN;
        public Dataset.GRMProject.SoilDepthDataTable mdtSoilDepthInfo;
        public Nullable<double> mConstSoilDepth;

        public cSetSoilDepth()
        {
        }

        public void SetValues(Dataset.GRMProject prjdb)
        {
            Dataset.GRMProject.ProjectSettingsRow row = (Dataset.GRMProject.ProjectSettingsRow)prjdb.ProjectSettings.Rows[0];
            if (mSoilDepthDataType.HasValue)
            {
                row.SoilDepthDataType = mSoilDepthDataType.Value.ToString();
                switch (mSoilDepthDataType.Value)
                {
                    case cGRM.FileOrConst.File:
                        {
                            row.SoilDepthFile = mGridSoilDepthFPN;
                            row.SoilDepthVATFile = mSoilDepthVATFPN;
                            row.SetConstantSoilDepthNull();
                            break;
                        }

                    case cGRM.FileOrConst.Constant:
                        {
                            row.SetSoilDepthFileNull();
                            row.SetSoilDepthVATFileNull();
                            row.ConstantSoilDepth = System.Convert.ToString(mConstSoilDepth.Value);
                            break;
                        }
                }
            }
            if (mSoilDepthDataType.Equals(cGRM.FileOrConst.File))
            {
                mdtSoilDepthInfo.AcceptChanges();
                foreach (DataRow r in mdtSoilDepthInfo.Rows)
                {
                    r.SetAdded();
                }
            }
        }

        public void GetValues(Dataset.GRMProject prjdb)
        {
            mSoilDepthDataType = default(cGRM.FileOrConst?);
            mSoilDepthVATFPN = null;
            mConstSoilDepth = default(Single?);
            Dataset.GRMProject.ProjectSettingsRow row = (Dataset.GRMProject.ProjectSettingsRow)prjdb.ProjectSettings.Rows[0];
            if (!row.IsSoilDepthDataTypeNull())
            {
                if (row.SoilDepthDataType.ToLower() == cGRM.FileOrConst.File.ToString().ToLower())
                {
                    mSoilDepthDataType = cGRM.FileOrConst.File;
                    mGridSoilDepthFPN = row.SoilDepthFile;
                    mSoilDepthVATFPN = row.SoilDepthVATFile;
                }
                else if (row.SoilDepthDataType.ToLower() == cGRM.FileOrConst.Constant.ToString().ToLower())
                {
                    mSoilDepthDataType = cGRM.FileOrConst.Constant;
                    double v = 0;
                    if (double.TryParse(row.ConstantSoilDepth, out v) == true) { mConstSoilDepth = v; }
                }
                else
                {
                    throw new InvalidOperationException();
                }
            }
            if (mSoilDepthDataType.Equals(cGRM.FileOrConst.File))
            {
                mdtSoilDepthInfo = new Dataset.GRMProject.SoilDepthDataTable();
                mdtSoilDepthInfo = prjdb.SoilDepth;
            }
            else
            { mdtSoilDepthInfo = null; }
        }

        public bool IsSet
        {
            get
            {
                return mSoilDepthDataType.HasValue;
            }
        }

        public static SoilDepthCode GetSoilDepthCode(string inName)
        {
            switch (inName.Trim())
            {
                case nameof(SoilDepthCode.D):
                    {
                        return SoilDepthCode.D;
                    }

                case nameof(SoilDepthCode.M):
                    {
                        return SoilDepthCode.M;
                    }

                case nameof(SoilDepthCode.S):
                    {
                        return SoilDepthCode.S;
                    }

                case nameof(SoilDepthCode.VD):
                    {
                        return SoilDepthCode.VD;
                    }

                case nameof(SoilDepthCode.VS):
                    {
                        return SoilDepthCode.VS;
                    }

                case nameof(SoilDepthCode.USER):
                    {
                        return SoilDepthCode.USER;
                    }

                case nameof(SoilDepthCode.CONSTV):
                    {
                        return SoilDepthCode.CONSTV;
                    }

                case nameof(SoilDepthCode.NULL):
                    {
                        return SoilDepthCode.NULL;
                    }

                default:
                    {
                        return default(SoilDepthCode);
                    }
            }
        }
    }
}
